src/serial.c

```diff
@@ -204,29 +217,17 @@ static void serial_handle_io(void *owner,
                              uint8_t size)
 {
     serial_dev_t *s = (serial_dev_t *) owner;
+    struct serial_dev_priv *priv = (struct serial_dev_priv *) s;
     void (*serial_op)(serial_dev_t *, uint16_t, void *) =
         is_write ? serial_out : serial_in;
 
+    pthread_mutex_lock(&priv->lock);
```

Comment:
It looks like the critical section only races with `serial_console`, which only access `rx_buf` there. Can we just lock smaller section to improve the concurrency?

Reply:
I've made the critical section smaller.
Not only `rx_buf`, but `iir`, `ier` and `lsr` can be accessed by the worker thread and the main thread. Some loads and stores to `iir` and `lsr` are modified to use atomic operations.